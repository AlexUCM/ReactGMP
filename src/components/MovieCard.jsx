import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { FallbackImage } from './FallbackImage';
import { CardMenu } from './CardMenu';
import styled from 'styled-components';

const Container = styled.div`
  display: flex;
  flex-direction: column;
  width: 100%;
  height: 100%;
  background-color: #fff;
`;

const Menu = styled.div`
  display: none;

  ${(props) =>
    props.isOpen &&
    `
    display: block;
  `}
`;

const Poster = styled.div`
  position: relative;
  padding-top: 148%;
  overflow: hidden;
  cursor: pointer;

  &:hover ${Menu} {
    display: block;
  }
`;

const Movie = styled.div`
  padding: 8px 8px 0 8px;
  display: flex;
  justify-content: space-between;
  background-color: #232323;
`;

const MovieInfo = styled.div`
  display: flex;
  flex-direction: column;
  overflow-x: hidden;
`;

const MovieTitle = styled.h2`
  margin: 0 0 4px 0;
  font-size: 18px;
  overflow-x: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  color: #ffffff7a;
`;

const MovieGenres = styled.p`
  font-size: 14px;
  overflow-x: hidden;
  text-overflow: ellipsis;
  text-transform: capitalize;
  white-space: nowrap;
  color: #555;
`;

const MovieDate = styled.div`
  font-size: 14px;
  margin: 5px 0 0 0;
  color: #ffffff7a;
`;

const DateString = styled.span`
  padding: 0 8px;
  border: 1px solid #555;
  border-radius: 5px;
`;

export const MovieCard = ({
  title,
  genres,
  date,
  poster,
  id,
  getMovie,
  handleMovieDet,
}) => {
  const [isOpen, setIsOpen] = useState(false);

  const toggleMenu = () => setIsOpen((state) => !state);

  const onOpenModal = (event) => {
    getMovie(id, event.target.innerHTML, true);
    setIsOpen((state) => !state);
  };

  return (
    <Container>
      <Poster onClick={() => handleMovieDet(id)}>
        <FallbackImage src={poster} />
        <Menu isOpen={isOpen}>
          <CardMenu
            isOpen={isOpen}
            onOpenModal={onOpenModal}
            toggleMenu={toggleMenu}
          />
        </Menu>
      </Poster>
      <Movie>
        <MovieInfo>
          <MovieTitle>{title}</MovieTitle>
          <MovieGenres>
            {genres.length === 2 ? genres.join(' & ') : genres.join(', ')}
          </MovieGenres>
        </MovieInfo>
        <MovieDate>
          <DateString>{date.slice(0, 4)}</DateString>
        </MovieDate>
      </Movie>
    </Container>
  );
};

MovieCard.propTypes = {
  title: PropTypes.string.isRequired,
  genres: PropTypes.arrayOf(PropTypes.string).isRequired,
  date: PropTypes.string.isRequired,
  poster: PropTypes.string,
  id: PropTypes.number.isRequired,
  getMovie: PropTypes.func.isRequired,
  handleMovieDet: PropTypes.func.isRequired,
};
