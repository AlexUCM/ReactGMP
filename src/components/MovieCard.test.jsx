import React from 'react';
import '@testing-library/jest-dom/extend-expect';
import { fireEvent, render } from '@testing-library/react';
import { MovieCard } from './MovieCard';

describe('MovieCard', () => {
  const getComponent = ({
    title = 'test title',
    genres = ['comedy'],
    date = '12/12/2020',
    poster = 'test src',
    id = 122,
    getMovie = jest.fn(),
    handleMovieDetails = jest.fn(),
  }) =>
    render(
      <MovieCard
        title={title}
        genres={genres}
        date={date}
        poster={poster}
        id={id}
        getMovie={getMovie}
        handleMovieDetails={handleMovieDetails}
      />
    );
  it('Should renders correctly', () => {
    expect(getComponent({}).asFragment()).toMatchSnapshot();
  });
  it('Should have class menu_opened', () => {
    const { getByTestId, getByText } = getComponent({});
    fireEvent.click(getByTestId('card-menu'));
    expect(getByTestId('menu')).toHaveStyle('display: block');
    fireEvent.click(getByText('Edit'));
    expect(getByTestId('menu')).toHaveStyle('display: none');
  });
  it('Sould to be called handleMovieDetails', () => {
    const handleMovieDetails = jest.fn();
    const { getByAltText } = getComponent({ handleMovieDetails });
    fireEvent.click(getByAltText('poster'));
    expect(handleMovieDetails).toBeCalled();
  });
});
