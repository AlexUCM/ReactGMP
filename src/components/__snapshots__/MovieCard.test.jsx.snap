// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieCard Should renders correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-gKsewC hYICKG"
  >
    <div
      class="sc-fubCfw ddvqw"
    >
      <img
        alt="poster"
        class="sc-bdfBwQ dnPwTH"
        height="16"
        src="test src"
        width="9"
      />
      <div
        class="sc-iBPRYJ jRTCDY"
        data-testid="menu"
      >
        <div
          class="sc-gsTCUz fTRmhn"
          data-testid="card-menu"
        >
          <div
            class="sc-dlfnbm cUkdsN"
          />
          <div
            class="sc-dlfnbm cUkdsN"
          />
          <div
            class="sc-dlfnbm cUkdsN"
          />
        </div>
      </div>
    </div>
    <div
      class="sc-pFZIQ eGSZAr"
    >
      <div
        class="sc-jrAGrp ezjpSe"
      >
        <h2
          class="sc-kEjbxe fgJrSU"
        >
          test title
        </h2>
        <p
          class="sc-iqHYGH fluhvY"
        >
          comedy
        </p>
      </div>
      <div
        class="sc-crrsfI hMBNEf"
      >
        <span
          class="sc-dQppl kCIcGn"
        >
          12/1
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
